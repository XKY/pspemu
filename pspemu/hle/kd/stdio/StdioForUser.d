module pspemu.hle.kd.stdio.StdioForUser; // kd/stdio.prx (sceStdio)

import pspemu.hle.ModuleNative;

class StdioForUser : ModuleNative {
	void initNids() {
		mixin(registerd!(0x172D316E, sceKernelStdin));
		mixin(registerd!(0xA6BAB2E9, sceKernelStdout));
		mixin(registerd!(0xF78BA90A, sceKernelStderr));
		mixin(registerd!(0x98220F3E, sceKernelStdoutReopen));
		mixin(registerd!(0xFB5380C5, sceKernelStderrReopen));
	}

	/**
	 * Function to get the current standard in file no
	 * 
	 * @return The stdin fileno
	 */
	SceUID sceKernelStdin() { return STDIN; }

	/**
	 * Function to get the current standard out file no
	 * 
	 * @return The stdout fileno
	 */
	SceUID sceKernelStdout() { return STDOUT; }

	/**
	 * Function to get the current standard err file no
	 * 
	 * @return The stderr fileno
	 */
	SceUID sceKernelStderr() { return STDERR; }

	/** 
	 * Function reopen the stdout file handle to a new file
	 *
	 * @param file - The file to open.
	 * @param flags - The open flags 
	 * @param mode - The file mode
	 * 
	 * @return < 0 on error.
	 */
	int sceKernelStdoutReopen(string file, int flags, SceMode mode) {
		unimplemented();
		return -1;
	}

	/** 
	 * Function reopen the stderr file handle to a new file
	 *
	 * @param file - The file to open.
	 * @param flags - The open flags 
	 * @param mode - The file mode
	 * 
	 * @return < 0 on error.
	 */
	int sceKernelStderrReopen(string file, int flags, SceMode mode) {
		unimplemented();
		return -1;
	}
}


enum : SceUID {
	STDIN  = -1,
	STDOUT = -2,
	STDERR = -3
}

static this() {
	mixin(ModuleNative.registerModule("StdioForUser"));
}