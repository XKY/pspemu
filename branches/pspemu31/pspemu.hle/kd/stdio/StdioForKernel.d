module pspemu.hle.kd.stdio.StdioForKernel; // kd/stdio.prx (sceStdio)

import pspemu.hle.ModuleNative;
import pspemu.hle.kd.stdio.StdioForUser;

class StdioForKernel : StdioForUser {
}

static this() {
	mixin(ModuleNative.registerModule("StdioForKernel"));
}