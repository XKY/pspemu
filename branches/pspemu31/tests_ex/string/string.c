//#pragma compile, "%PSPSDK%/bin/psp-gcc" -I. -I"%PSPSDK%/psp/sdk/include" -L. -L"%PSPSDK%/psp/sdk/lib" -D_PSP_FW_VERSION=150 -Wall -g -O0 string.c ../common/emits.c -lpspsdk -lc -lpspuser -lpspkernel -o test.elf
//#pragma compile, "%PSPSDK%/bin/psp-fixup-imports" string.elf

#include <pspkernel.h>
#include <pspthreadman.h>
#include <pspdebug.h>
#include <stdio.h>
#include <string.h>
#include <../emits.h>

PSP_MODULE_INFO("THREAD TEST", 0, 1, 1);
PSP_MAIN_THREAD_ATTR(THREAD_ATTR_USER | THREAD_ATTR_VFPU);

int main(int argc, char** argv) {
	char buffer[128];
	
	pspDebugScreenInit();
	
	sprintf(buffer, "%d", (u32)0);
	emitString(buffer);
	pspDebugScreenPrintf("%s\n", buffer);
	
	sprintf(buffer, "%d", (u32)100000);
	emitString(buffer);
	pspDebugScreenPrintf("%s\n", buffer);
	
	sprintf(buffer, "%lld", (u64)100000);
	emitString(buffer);
	pspDebugScreenPrintf("%s\n", buffer);

	sceKernelExitGame();
	return 0;
}
