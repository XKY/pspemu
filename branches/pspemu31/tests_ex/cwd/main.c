/*
 * PSP Software Development Kit - http://www.pspdev.org
 * -----------------------------------------------------------------------
 * Licensed under the BSD license, see LICENSE in PSPSDK root for details.
 *
 * main.c - Basic sample to demonstrate some fileio functionality.
 *
 * Copyright (c) 2005 Marcus R. Brown <mrbrown@ocgnet.org>
 * Copyright (c) 2005 James Forshaw <tyranid@gmail.com>
 * Copyright (c) 2005 John Kelley <ps2dev@kelley.ca>
 * Copyright (c) 2005 Jim Paris <jim@jtan.com>
 *
 * $Id: main.c 1175 2005-10-20 15:41:33Z chip $
 */
#include <pspkernel.h>
#include <pspctrl.h>
#include <pspdebug.h>
#include <pspdisplay.h>
#include <psptypes.h>
#include <pspiofilemgr.h>

#include <stdio.h>
#include <string.h>
#include <sys/types.h>
#include <sys/unistd.h>

PSP_MODULE_INFO("CwdTest", 0, 1, 1);

void try(const char *dest)
{
	char buf[MAXPATHLEN];

	Kprintf("%16s --> ", dest);
	if(chdir(dest) < 0) {
		Kprintf("(chdir error)\n");
	} else {
		Kprintf("%s\n", getcwd(buf, MAXPATHLEN) ?: "(getcwd error)");
	}
}

int main(int argc, char *argv[])
{
	int n;
	char buf[MAXPATHLEN];

	Kprintf("Working Directory Examples\n");
	Kprintf("Arguments: %d\n", argc);
	for (n = 0; n < argc; n++) {
		Kprintf("Argument[%d]: '%s'\n", n, argv[n]);
	}
	Kprintf("Initial dir: %s\n\n", getcwd(buf, MAXPATHLEN) ?: "(error)");

	Kprintf("%16s --> %s\n", "chdir() attempt", "resulting getcwd()");
	Kprintf("%16s --> %s\n", "---------------", "------------------");
	try("");		   /* empty string                */
	try("hello");		   /* nonexistent path            */
	try("..");		   /* parent dir                  */
	try("../SAVEDATA");	   /* parent dir and subdir       */
	try("../..");		   /* multiple parents            */
	try(".");		   /* current dir                 */
	try("./././//PSP");        /* current dirs, extra slashes */
	try("/PSP/./GAME");	   /* absolute with no drive      */
	try("/");                  /* root with no drive          */
	try("ms0:/PSP/GAME");      /* absolute with drive         */
	try("flash0:/");           /* different drive             */
	try("ms0:/PSP/../PSP/");   /* mixed                       */

	Kprintf("\nAll done!\n");

	return 0;
}
