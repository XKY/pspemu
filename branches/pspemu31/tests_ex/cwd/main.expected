Working Directory Examples
Arguments: 1
Argument[0]: 'ms0:/PSP/GAME/virtual/main.elf'
Initial dir: ms0:/PSP/GAME/virtual

 chdir() attempt --> resulting getcwd()
 --------------- --> ------------------
                 --> ms0:/PSP/GAME/virtual
           hello --> ms0:/PSP/GAME/virtual/hello
              .. --> ms0:/PSP/GAME/virtual
     ../SAVEDATA --> ms0:/PSP/GAME/SAVEDATA
           ../.. --> ms0:/PSP
               . --> ms0:/PSP
     ./././//PSP --> ms0:/PSP/PSP
     /PSP/./GAME --> ms0:/PSP/GAME
               / --> ms0:
   ms0:/PSP/GAME --> ms0:/PSP/GAME
        flash0:/ --> flash0:
ms0:/PSP/../PSP/ --> ms0:/PSP

All done!